### PR TITLE
feat(#400): allow lookbehind tuple

### LINKED ISSUE #400: feat: allow lookbehind tuple

### LINKED ISSUE #402: feat: allow eager polarsframes as input

### LINKED ISSUE #403: refactor: rename slicedframe to timemaskedframe

### LINKED ISSUE #423: perf: multiprocessing

### PR DESCRIPTION
**Pull request by HLasse**
feat(#400): allow lookbehind tuple

Fixes #400

### CODE REVIEW

**Comment by HLasse** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #443** <a href="https://app.graphite.dev/github/pr/Aarhus-Psychiatry-Research/timeseriesflattener/443?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/Aarhus-Psychiatry-Research/timeseriesflattener/443?utm_source=stack-comment).

**Comment by HLasse** [comment]
@MartinBernstorff unsure what has happened here :O

File: src/timeseriesflattenerv2/feature_specs.py [hunk 1]
```diff
@@ -137,23 +137,65 @@ def __call__(self, column_name: str) -> pl.Expr:
...
+            _lookdistance_to_normalised_lookperiod(lookdistance=lookdistance, direction="behind")
+            for lookdistance in self.lookbehind_distances
+        ]
+
 
 @dataclass(frozen=True)
 class OutcomeSpec:
     value_frame: ValueFrame
-    lookahead_distances: Sequence[LookDistance]
+    lookahead_distances: Sequence[LookDistance | tuple[LookDistance, LookDistance]]
```
Comment (MartinBernstorff) [inline src/timeseriesflattenerv2/feature_specs.py:187]:
```suggestion
    lookahead_distances: Sequence[LookDistance| | Sequence[tuple[LookDistance, LookDistance]]
```
I don't think we should allow both types at once?

File: src/timeseriesflattenerv2/feature_specs.py [hunk 2]
```diff
@@ -137,23 +137,65 @@ def __call__(self, column_name: str) -> pl.Expr:
...
     aggregators: Sequence[Aggregator]
     fallback: ValueType
     column_prefix: str = "outc"
 
+    @property
+    def normalised_lookperiod(self) -> Sequence[LookPeriod]:
+        return [
+            _lookdistance_to_normalised_lookperiod(lookdistance=lookdistance, direction="ahead")
+            for lookdistance in self.lookahead_distances
+        ]
```
Comment (MartinBernstorff) [inline src/timeseriesflattenerv2/feature_specs.py:197]: Should this maybe happen in `__post_init__`?

Then it is never in an invalid state. And we can check that the values are reasonable.

**Review by MartinBernstorff — APPROVED** [review]
This looks good!

I'd prefer a design where we go to the absolute scale immediately (e.g. a predictor is represented by a .window_start that is a negative datetime, and a .window_end that is negative or 0). I think that would make the code easier to read, and push out complexity (fewer `abs()` and comparisons).

Not a hard requirement, though ❤️

File: src/timeseriesflattenerv2/feature_specs.py [hunk 3]
```diff
@@ -137,23 +137,65 @@ def __call__(self, column_name: str) -> pl.Expr:
         ...
 
 
+@dataclass(frozen=True)
+class LookPeriod:
+    closest_to_prediction_time: LookDistance
+    furthest_from_prediction_time: LookDistance
+
+    def __post_init__(self):
+        if abs(self.closest_to_prediction_time) > abs(self.furthest_from_prediction_time):
```
Comment (MartinBernstorff) [inline src/timeseriesflattenerv2/feature_specs.py:146]:
```suggestion
        if abs(self.closest_to_prediction_time) >= abs(self.furthest_from_prediction_time):
```

**Comment by MartinBernstorff** [comment]
I think we should decide on this before we finalise #444.

File: src/timeseriesflattenerv2/feature_specs.py (hunk 1 above)
Comment (MartinBernstorff) [reply src/timeseriesflattenerv2/feature_specs.py]: I'd like to support both
* Sequence[LookDistance] 
and
* Sequence[tuple[LookDistance, LookDistance]]

But not both at once - I'd imagine that would be due to a typo! But maybe not? Not a strong preference.

File: src/timeseriesflattenerv2/test_flattener.py [hunk 4]
```diff
@@ -26,7 +32,7 @@
         ),
         value_col_name="FakeValueColName",
     ),
-    lookbehind_distances=[dt.timedelta(days=1)],
+    lookbehind_distances=[(dt.timedelta(days=0), dt.timedelta(days=1))],
```
Comment (MartinBernstorff) [inline src/timeseriesflattenerv2/test_flattener.py:35]: Måske en god idé at vi stadig understøtter `Sequence[LookDistance]` her? Så kan vi re-caste i post-init?

File: src/timeseriesflattenerv2/feature_specs.py [hunk 5]
```diff
@@ -138,21 +138,57 @@ def __call__(self, column_name: str) -> pl.Expr:
...
+            for lookdistance in self.lookbehind_distances
+        ]
 
-@dataclass(frozen=True)
+
+@dataclass()
 class OutcomeSpec:
     value_frame: ValueFrame
-    lookahead_distances: Sequence[LookDistance]
+    lookahead_distances: Sequence[tuple[LookDistance, LookDistance]]
```
Comment (MartinBernstorff) [inline src/timeseriesflattenerv2/feature_specs.py:181]:
```suggestion
    lookahead_distances: InitVar[Sequence[tuple[LookDistance, LookDistance]]]
```

File: src/timeseriesflattenerv2/feature_specs.py [hunk 6]
```diff
@@ -138,21 +138,57 @@ def __call__(self, column_name: str) -> pl.Expr:
...
     aggregators: Sequence[Aggregator]
     fallback: ValueType
     column_prefix: str = "outc"
+    normalised_lookperiod: Sequence[LookPeriod] = field(init=False)
+
+    def __post_init__(self):
```
Comment (MartinBernstorff) [inline src/timeseriesflattenerv2/feature_specs.py:187]:
```suggestion
    def __post_init__(self, lookahead_distances: ...):
```

File: src/timeseriesflattenerv2/_process_spec.py [hunk 7]
```diff
@@ -61,21 +49,22 @@ def _null_values_outside_lookwindow(
...
+    new_colname_prefix = f"{column_prefix}_{value_col_name}_within_"
 
     # The predictor case
     if is_lookbehind:
-        after_lookbehind_start = lookdistance <= timedelta_col
-        before_prediction_time = timedelta_col <= dt.timedelta(0)
+        new_colname = (
+            new_colname_prefix + f"{abs(lookperiod.last.days)}_to_{abs(lookperiod.first.days)}_days"
+        )
+        after_lookbehind_start = timedelta_col >= lookperiod.first
```
Comment (MartinBernstorff) [inline src/timeseriesflattenerv2/_process_spec.py:66]:
```suggestion
        after_lookbehind_start = lookperiod.first <= timedelta_col
```

Just a small readability preference ❤️

File: src/timeseriesflattenerv2/test_specs.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+MockValueFrame = ValueFrame(
+    init_df=pl.LazyFrame({"value": [1], "timestamp": ["2021-01-01"], "entity_id": [1]}),
+    value_col_name="value",
+)
+
+
+def test_predictor_spec_post_init():
+    lookdistance_start = dt.timedelta(days=1)
+    lookdistance_days = dt.timedelta(days=10)
```
Comment (MartinBernstorff) [inline src/timeseriesflattenerv2/test_specs.py]: Naming here a bit unintuitive. Perhaps lookdistance start and end?

**Review by MartinBernstorff — COMMENTED** [review]
Looks great Lasse! I find it to  be much more readable. A few comments re: interface, then I think it's good to go!

File: src/timeseriesflattenerv2/feature_specs.py (hunk 1 above)
Comment (HLasse) [reply src/timeseriesflattenerv2/feature_specs.py]: I see, completely misunderstood your suggested change! Hm, one might want to use the shorthand for the first interval and then make the rest with tuples (could myself doing that at least), so I vote `Sequence[LookDistance | tuple[LookDistance, LookDistance]]`

**Pull request merged**